### PR TITLE
[MAINT] Drop Python 3.4 support

### PR DESCRIPTION
**Pull request by hameerabbasi**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -68,7 +57,15 @@ matrix:
        - PYTHONOPTIMIZE=2
        - USE_ASV=1
     - python: 3.5
-      env: NPY_RELAXED_STRIDES_CHECKING=0
+      env: USE_DEBUG=1 NPY_RELAXED_STRIDES_CHECKING=0
```
Comment (hameerabbasi) [inline .travis.yml]: I moved the `USE_DEBUG` flag and related packages from Python 3.4 to here.

**Review by mhvk — COMMENTED** [review]
Looks good to me, though it might be good for @charris to check that the whole still makes sense (to me, there seem to be too many test cases, though perhaps that is better addressed in another PR)

**Comment by hameerabbasi** [comment]
It seems the Travis failure is present on `master` as well and is a Travis issue. Can someone double-check?

**Comment by charris** [comment]
We might look to drop some of the special test cases, relaxed strides, for instance.

**Comment by charris** [comment]
I've seen similar travis failures on numpy-wheels. They seem intermittent.

**Comment by hameerabbasi** [comment]
From the log, it seems the "new" Python 3.5 test still tests Python 3.4.3. It appears the debug packages themselves are for Python 3.4. I've changed to use `python3.5-dev` and `python3.5-dbg`.

File: doc/release/1.16.0-notes.rst [hunk 2]
```diff
@@ -6,6 +6,8 @@ NumPy 1.16.0 Release Notes
 Highlights
 ==========
 
+* Python 3.4 support has officially been dropped.
```
Comment (eric-wieser) [inline doc/release/1.16.0-notes.rst]: Not sure we need the word "official" in the release notes which are distributed "officially" anyway.

**Comment by hameerabbasi** [comment]
It appears the Python 3.5 dev/debug packages aren't available in Trusty (Ubuntu 14.04), and Travis has no plans to upgrade to Xenial (16.04) or later. Any possible alternatives?

**Comment by eric-wieser** [comment]
To be clear, 1.16 is also the version we drop 2.7 support? If not, I see no benefit to dropping 3.4 support yet.

**Comment by charris** [comment]
@eric-wieser No, we drop 2.7 in 1.17. Note that Python 3.7 is due at the end of the month and I don't see the point of producing wheels for 3.4. I doubt that we will do anything that breaks on 3.4, but there is no reason to promise that it will stay working.

**Comment by hameerabbasi** [comment]
If the `USE_DEBUG=1` flag isn't critical, I could just drop it. IMHO the debugging should be done locally if the test fails.

**Comment by mhvk** [comment]
Do keep the `USE_DEBUG` - I've found lots of problems using it in C code (mostly with forgetting to DECREF references, but also because assert statements in C code are actually being executed) - and to run it locally needs installation of a debug environment, which is not really trivial.

But more generally it would be good if the `.travis.yml` had sufficient comments to know why a given setup was being tested...

**Comment by hameerabbasi** [comment]
It seems the [`python3-dev`](https://packages.ubuntu.com/trusty/python3-dev) and [`python3-dbg`](https://packages.ubuntu.com/trusty/python3-dbg) packages in Ubuntu 14.04 (trusty) (in the default Travis build environment) point to Python 3.4.3. [`python3.5-dev`](https://packages.ubuntu.com/xenial/python3.5-dev) and [`python3.5-dbg`](https://packages.ubuntu.com/xenial/python3.5-dbg) are only available in Ubuntu 16.04 (xenial), and no other version. Travis only offers trusty, with no plans to move to xenial or later (xref travis-ci/travis-ci#5821).

I'm unsure how to proceed. I guess I could just get Python 3.5 dev and debug from source and build it on the container... But that'd require complicating the CI process. I'm unsure how to proceed.

Thoughts?

**Comment by mattip** [comment]
@matthew-brett any suggestions?

**Comment by charris** [comment]
The suggested solution is to use docker, however, apparently xenial was in the works last Dec and there is also a new `Group` option for the yml file. See https://blog.travis-ci.com/2017-12-01-new-update-schedule-for-linux-build-images. I'm sure there is more information out there, and the easiest thing todo might be to ask the travisci folks.

**Comment by rgommers** [comment]
Remark: we don't have to drop Python 3.4, if it's more work than a few hours, then leave it in for now. It's not like we're gaining much from dropping it before we drop 2.7

**Comment by charris** [comment]
Yes, there is no rush, and travisCI may solve our problem in the interim. The
```
group: travis_latest
```
option does look intriguing.

**Comment by matthew-brett** [comment]
Nasty.  I've just checked, I believe you could use pyenv.  This is in a bare Trusty container:

```
sudo apt-get update
sudo apt-get install -y make build-essential libssl-dev zlib1g-dev libbz2-dev libreadline-dev libsqlite3-dev wget curl llvm libncurses5-dev libncursesw5-dev xz-utils tk-dev libffi-dev
sudo apt-get install -y git
curl -L https://github.com/pyenv/pyenv-installer/raw/master/bin/pyenv-installer | bash
export PATH="$HOME/.pyenv/bin:$PATH"
eval "$(pyenv init -)"
eval "$(pyenv virtualenv-init -)"
pyenv install --debug 3.5.5
pyenv global 3.5.5-debug
pip install --no-binary cython cython
```

It takes about 5 minutes on a reasonably fast machine.

**Comment by mattip** [comment]
Should we close this, and revisit when we are closer to dropping 2.7?

**Comment by hameerabbasi** [comment]
+1 on closing from me, I'll wait on consensus though.

**Comment by charris** [comment]
Let's revisit in the Fall.

**Comment by charris** [comment]
Thanks @hameerabbasi for the try.

**Pull request closed without merging**